File: all-submissions/2392-successful-pairs-of-spells-and-potions/solution.java
class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        if (spells.length == 0) return new int[]{0};
        int[] answer = new int[spells.length];
        int count;
        Arrays.sort(potions);
        for (int i = 0; i<spells.length;i++) {
            // if (potions[potions.length-1] * spells[i] < success) {
            //     count = 0;
            // } else {
                count = binarySearch(spells[i], potions, success);
            // }
            answer[i] = count;
        }
        return answer;
    }

    private int binarySearch(int spell, int[] potions, long success) {
        int left = 0;
        int right = potions.length-1;
        while(left <=right) {
            int mid = left + (right-left) /2;
            if (potions[mid] * (long) spell < success) {
                left = mid +1;
            }
            else if (potions[mid] * (long) spell >= success) {
                right = mid-1;
            }
        }
        return potions.length - left;
    }
}

File: all-submissions/2392-successful-pairs-of-spells-and-potions/solution.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        res = []
        for i in range(len(spells)):
            l = 0
            r = len(potions) - 1
            count = 0
            spell = spells[i]
            while l <= r:
                mid = (l+r)//2
                prod = spell * potions[mid]
                if prod >= success:
                    count += r-mid+1
                    r = mid-1
                else:
                    l = mid+1
            res.append(count)
        return res

